client/blocks/payment-request/hooks.js

```diff
@@ -34,33 +36,31 @@ export const usePaymentRequest = ( stripe, needsShipping, billing ) => {
 	//   a) Stripe object is loaded; and
 	//   b) There is no payment request created already.
 	useEffect( () => {
-		// Do nothing if Stripe object isn't loaded or paymentRequest already exists.
-		if ( ! stripe || paymentRequest ) {
+		if ( ! stripe ) {
 			return;
 		}
-
-		getCartDetails().then( ( cart ) => {
+		const getPaymentRequest = async () => {
+			const cart = await getCartDetails();
 			const pr = createPaymentRequestUsingCart( stripe, cart );
+			const result = await pr.canMakePayment();
 
-			pr.canMakePayment().then( ( result ) => {
-				if ( result ) {
-					setPaymentRequest( pr );
+			if ( result ) {
+				setPaymentRequest( pr );
+				setPaymentRequestType( () => {
 					if ( result.applePay ) {
-						setPaymentRequestType( 'apple_pay' );
-					} else if ( result.googlePay ) {
-						setPaymentRequestType( 'google_pay' );
-					} else {
-						setPaymentRequestType( 'payment_request_api' );
+						return 'apple_pay';
 					}
-				}
-			} );
-		} );
-	}, [ paymentRequest, stripe ] );
-
-	// Reset the payment request if the need for shipping changes.
-	useEffect( () => {
+					if ( result.googlePay ) {
+						return 'google_pay';
+					}
+					return 'payment_request_api';
+				} );
+			}
+		};
 		setPaymentRequest( null );
```

Comment:
I don't know, I just re-implemented what was here before. It will be null inbetween the payment request being invalidated then recreated.

Reply:
Yes, the `null` is necessary so we can recreate the the payment request when the shipping needs in the cart change.

For example; you have a product that needs shipping and a product that doesn't need shipping in your cart. Without this reset the payment request will insist on a shipping address even if you remove the product that needs shipping from the cart.

This happens because we can't update the `needsShipping` prop on the payment request; we must create a new payment request for that to work.

Reply:
With @mikejolley's change here, I think it will rerun the one effect (and so still recreate the payment request) when those same dependencies change. Because of that, It doesn't look to me like setting it to `null` is necessary anymore – except to reset it in the logic above, if `! canMakePayment`.

Reply:
Ah, I didn't notice that.

I initially split them up because we only want to invalidate the payment request when specific variables change, and it's not necessary to recreate the whole payment request every time — which is quite expensive because of the network request involved to get the cart.

I haven't looked into whether joining the 2 effects increases the number of times we reset the payment request, but I think we should verify that before joining them.

Aside from that, I think you're right Paul; we can remove the `setPaymentRequest( null )` call if we decide to join the effects like this 🙂

Reply:
Done in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1603/commits/29cb4e470aa7e6570458046177672aa92ea2d95e